walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,171 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Arrays;
+import java.util.Set;
+import java.util.HashSet;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+
+        List<String> allAttendees = Stream.concat(request.getAttendees().stream(), request.getOptionalAttendees().stream())
+                                        .collect(Collectors.toList());
+        ArrayList<Event> validEventsAllAttendees = ignoreEventsWithoutRequestAttendees(events, allAttendees);
+
+        //if the duration of the meeting is over a day --> no available times
+        if(request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+            return new ArrayList<TimeRange>();
+        }
+
+        //if there are no events or no attendees, the entire day is available
+        if(validEventsAllAttendees.size() == 0 || allAttendees.size() == 0) {
+            return Arrays.asList(TimeRange.WHOLE_DAY);
+        }
+
+        ArrayList<TimeRange> availableTimesAllAttendees = getAvailableTimes(validEventsAllAttendees, request);
+
+        //if there was one or more times available with optional attendees, return those times
+        if(availableTimesAllAttendees.size() > 0) {
+            return availableTimesAllAttendees;
+        }
+
+        //if there is no time with both mandatory and optional, just look at mandatory attendees
+        ArrayList<Event> validEventsMandatoryAttendees = ignoreEventsWithoutRequestAttendees(events, request.getAttendees());
+        
+        //if we got to this statement, that means there were only optional attendees and no mandatory attendees,
+        //so there is no time available
+        if(request.getAttendees().size() == 0) {
```

Comment:
This should probably go before the previous statement where you get the validEventsMandatoryAttendees